File: tools/build-anchor.mjs
#!/usr/bin/env node
/**
 * Livello 2 — figma.json: l'ANCORAGGIO.
 *
 * È l'input dell'estrazione, non il suo output: serve per sapere cosa andare a
 * prendere. Per questo è un file a sé e non un blocco dentro contract.json —
 * cambia solo quando un nodo si sposta, mentre il contratto cambia a ogni
 * pubblicazione.
 *
 * Cerca anche i FRAME DI DOCUMENTAZIONE nel file: è la misura di PA-19, che
 * non era mai stata fatta.
 *
 *   node build-anchor.mjs <slug> <nodeId>
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT, TOKEN } from './_env.mjs';

const OUT = `${ROOT}/data/contracts`;

/* Vedi nota in figma-files.json: gli id sono condivisi fra i due file, il ds
   deve essere esplicito. */
const CFG = JSON.parse(fs.readFileSync(`${ROOT}/config/figma-files.json`, 'utf8'));

const slug = process.argv[2] ?? 'button';
const nodeId = process.argv[3] ?? '5473-10855';
const ds = process.argv[4] ?? CFG.default;

if (!CFG.files[ds]) {
  console.error(`\n  ✖ ds sconosciuto: "${ds}" — attesi: ${Object.keys(CFG.files).join(', ')}\n`);
  process.exit(1);
}
const FILE = CFG.files[ds].fileKey;

/* Su B2C un componente ha un nodo per piattaforma: si leggono dal metadata,
   se esiste. Su B2B la piattaforma è una sola. */
let platformNodes = { default: nodeId };
try {
  const m = JSON.parse(fs.readFileSync(`${ROOT}/components/${slug}/docs/metadata.json`, 'utf8'));
  if (m.figmaNodeIds && ds === 'b2c') {
    const { propertyTable, ...plat } = m.figmaNodeIds;
    if (Object.keys(plat).length) platformNodes = plat;
  }
} catch { /* nessun metadata: si resta al nodo passato a riga di comando */ }

const api = async u => {
  const r = await fetch(`https://api.figma.com/v1/${u}`, { headers: { 'X-Figma-Token': TOKEN } });
  if (!r.ok) throw new Error(`HTTP ${r.status} — ${u}`);
  return r.json();
};

/* ── il nodo ──────────────────────────────────────────────────────────────── */
const nodes = await api(`files/${FILE}/nodes?ids=${nodeId}`);
const entry = Object.values(nodes.nodes)[0];
const doc = entry.document;

/* ── i frame di documentazione esistono? (PA-19) ─────────────────────────── */
const DOC_RE = /purpose|usage|behavior|behaviour|anatomy|do\s*&|don'?t|guideline|spec|documentation|linee guida/i;
const file = await api(`files/${FILE}?depth=3`);
const found = [];
const canvases = file.document.children ?? [];
(function walk(nodes, page, depth = 0) {
  for (const n of nodes ?? []) {
    if (depth > 0 && DOC_RE.test(n.name ?? '')) found.push({ page, id: n.id, name: n.name, type: n.type });
    walk(n.children, page, depth + 1);
  }
})(canvases.flatMap(c => (c.children ?? []).map(x => ({ ...x, _page: c.name }))), null, 1);

// rifatto in modo leggibile: per pagina
const byPage = [];
for (const c of canvases) {
  const hits = [];
  (function w(ns) { for (const n of ns ?? []) { if (DOC_RE.test(n.name ?? '')) hits.push({ id: n.id, name: n.name, type: n.type }); w(n.children); } })(c.children);
  byPage.push({ page: c.name, id: c.id, children: (c.children ?? []).length, docFrames: hits });
}

/* ── lo stato, dal marcatore nel nome della pagina ───────────────────────── */
const STATUS = JSON.parse(fs.readFileSync(
  `${ROOT}/config/figma-status.json`, 'utf8'));
const EMOJI = /[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{2B00}-\u{2BFF}\u{FE0F}\u{25A0}-\u{25FF}]/gu;

// la pagina che contiene il nodo. `depth=3` può non arrivare al nodo, quindi
// se la ricerca nell'albero fallisce si rilegge il file più in profondità.
const wanted = nodeId.replace('-', ':');
const contains = c => { let hit = false; (function w(ns) { for (const n of ns ?? []) { if (n.id === wanted) hit = true; w(n.children); } })(c.children); return hit; };
let ownerPage = canvases.find(contains);
if (!ownerPage) {
  const deep = await api(`files/${FILE}?depth=5`);
  ownerPage = (deep.document.children ?? []).find(contains);
}
const marker = ownerPage ? [...new Set((ownerPage.name.match(EMOJI) ?? []))].join('') : null;
const map = STATUS[/DS B2B/.test(file.name) ? 'b2b' : 'b2c'] ?? {};

const anchor = {
  id: `ds.${slug}`,
  fileKey: FILE,
  fileName: file.name,
  ds,
  nodes: platformNodes,                    // b2c: un nodo per piattaforma · b2b: uno solo
  nodeType: doc.type,
  nodeName: doc.name,
  variantCount: (doc.children ?? []).length,
  parts: {},                               // sotto-nodi, se e quando servono
  page: ownerPage ? {
    id: ownerPage.id,
    name: ownerPage.name,
    label: ownerPage.name.replace(EMOJI, '').trim(),
    marker: marker || null,
    status: marker ? (map[marker] ?? 'unknown') : null,
  } : null,
  documentation: byPage.filter(p => p.docFrames.length)
    .map(p => ({ page: p.page, frames: p.docFrames })),
  lastModified: file.lastModified,
};

fs.mkdirSync(OUT, { recursive: true });
const dir = path.join(OUT, slug);
fs.mkdirSync(dir, { recursive: true });
const out = path.join(dir, 'figma.json');
fs.writeFileSync(out, JSON.stringify(anchor, null, 2) + '\n');

const L = '─'.repeat(70);
console.log(`\n${L}\n  ANCORAGGIO — ${doc.name}\n${L}`);
console.log(`\n   file      "${file.name}"  ${FILE}`);
console.log(`   nodo      ${nodeId}  ${doc.type}  ·  ${anchor.variantCount} varianti`);
console.log(`   modificato ${file.lastModified}`);
console.log(`\n▸ PAGINE DEL FILE (${canvases.length})`);
for (const p of byPage) console.log(`   ${String(p.children).padStart(4)} figli  ${p.page}${p.docFrames.length ? `   ← ${p.docFrames.length} frame di doc` : ''}`);
const total = byPage.reduce((s, p) => s + p.docFrames.length, 0);
console.log(`\n▸ FRAME DI DOCUMENTAZIONE  (misura di PA-19)`);
console.log(`   trovati: ${total}`);
for (const p of byPage.filter(x => x.docFrames.length)) {
  console.log(`   — ${p.page}`);
  p.docFrames.slice(0, 8).forEach(f => console.log(`        ${f.type.padEnd(14)} ${f.name}`));
}
if (!total) console.log(`   ⚠️  nessun frame con nomi tipo Purpose / Usage / Behavior / Anatomy / Do & Don't`);
console.log(`\n✅ scritto data/contracts/${slug}/figma.json\n${L}\n`);

File: tools/build_anchor.py
#!/usr/bin/env python3
"""
Livello 2 — figma.json: l'ANCORAGGIO.

E' l'input dell'estrazione, non il suo output: serve per sapere cosa andare a
prendere. Per questo e' un file a se' e non un blocco dentro contract.json:
cambia solo quando un nodo si sposta, mentre il contratto cambia a ogni
pubblicazione.

Censisce anche i frame di documentazione presenti nel file.

⚠️ La ricerca dei frame usa `depth=3`, che e' una trappola nota: la
documentazione piu' annidata non viene vista, e il risultato sembra «nessuna
documentazione» invece di «non ho cercato abbastanza». Il censimento vero lo fa
`classify_pages.py`; qui il dato resta come indicazione, non come misura.

  python tools/build_anchor.py <slug> <nodeId> [ds] [--forza]
"""
import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from _figma import ROOT, Figma, scrivi_json  # noqa: E402
from _guardia import verifica_nodo  # noqa: E402

USCITA = ROOT / "data" / "contracts"

DOC_RE = re.compile(r"purpose|usage|behavior|behaviour|anatomy|do\s*&|don'?t|guideline|spec|documentation|linee guida", re.I)
EMOJI = re.compile("[\U0001F300-\U0001FAFF\u2600-\u27BF\u2B00-\u2BFF\uFE0F\u25A0-\u25FF]")


def main(argv):
    resto = [a for a in argv if not a.startswith("--")]
    slug = resto[0] if resto else "button"
    node_id = resto[1] if len(resto) > 1 else "5473-10855"
    cfg = json.loads((ROOT / "config" / "figma-files.json").read_text(encoding="utf-8"))
    ds = resto[2] if len(resto) > 2 else cfg["default"]
    if ds not in cfg["files"]:
        sys.exit(f'\n  ✖ ds sconosciuto: "{ds}" — attesi: {", ".join(cfg["files"])}\n')
    chiave_file = cfg["files"][ds]["fileKey"]

    # Su B2C un componente ha un nodo per piattaforma: si leggono dal metadata
    # della knowledge base, se esiste. Su B2B la piattaforma e' una sola.
    nodi_piattaforma = {"default": node_id}
    try:
        m = json.loads((Path("C:/Users/simon/Downloads/ds-cross-app") / "components" / slug
                        / "docs" / "metadata.json").read_text(encoding="utf-8"))
        if m.get("figmaNodeIds") and ds == "b2c":
            plat = {k: v for k, v in m["figmaNodeIds"].items() if k != "propertyTable"}
            if plat:
                nodi_piattaforma = plat
    except (OSError, json.JSONDecodeError):
        pass

    f = Figma()
    nodi = f.get(f"files/{chiave_file}/nodes?ids={node_id}")
    doc = list(nodi["nodes"].values())[0]["document"]
    verifica_nodo(slug, doc.get("name"), ds=ds, node_id=node_id, forza="--forza" in argv)

    file_meta = f.get(f"files/{chiave_file}?depth=3")
    canvases = (file_meta.get("document") or {}).get("children") or []

    per_pagina = []
    for c in canvases:
        colpi = []

        def scendi(ns):
            for n in ns or []:
                if DOC_RE.search(n.get("name") or ""):
                    colpi.append({"id": n.get("id"), "name": n.get("name"), "type": n.get("type")})
                scendi(n.get("children"))

        scendi(c.get("children"))
        per_pagina.append({"page": c.get("name"), "id": c.get("id"),
                           "children": len(c.get("children") or []), "docFrames": colpi})

    # ── lo stato, dal marcatore nel nome della pagina ────────────────────────
    stato_cfg = json.loads((ROOT / "config" / "figma-status.json").read_text(encoding="utf-8"))
    voluto = node_id.replace("-", ":")

    def contiene(c):
        trovato = [False]

        def w(ns):
            for n in ns or []:
                if n.get("id") == voluto:
                    trovato[0] = True
                w(n.get("children"))

        w(c.get("children"))
        return trovato[0]

    pagina = next((c for c in canvases if contiene(c)), None)
    if pagina is None:
        # depth=3 puo' non arrivare al nodo: si rilegge piu' a fondo
        profondo = f.get(f"files/{chiave_file}?depth=5")
        pagina = next((c for c in (profondo.get("document") or {}).get("children") or []
                       if contiene(c)), None)

    marcatore = ""
    if pagina:
        visti = []
        for m in EMOJI.findall(pagina.get("name") or ""):
            if m not in visti:
                visti.append(m)
        marcatore = "".join(visti)
    mappa = stato_cfg.get("b2b" if re.search(r"DS B2B", file_meta.get("name") or "") else "b2c") or {}

    ancoraggio = {
        "id": f"ds.{slug}",
        "fileKey": chiave_file,
        "fileName": file_meta.get("name"),
        "ds": ds,
        "nodes": nodi_piattaforma,
        "nodeType": doc.get("type"),
        "nodeName": doc.get("name"),
        "variantCount": len(doc.get("children") or []),
        "parts": {},
        "page": ({
            "id": pagina.get("id"),
            "name": pagina.get("name"),
            "label": EMOJI.sub("", pagina.get("name") or "").strip(),
            "marker": marcatore or None,
            "status": (mappa.get(marcatore, "unknown") if marcatore else None),
        } if pagina else None),
        "documentation": [{"page": p["page"], "frames": p["docFrames"]}
                          for p in per_pagina if p["docFrames"]],
        "lastModified": file_meta.get("lastModified"),
    }

    cartella = USCITA / slug
    cartella.mkdir(parents=True, exist_ok=True)
    scrivi_json(cartella / "figma.json", ancoraggio)

    L = "─" * 70
    print(f"\n{L}\n  ANCORAGGIO — {doc.get('name')}\n{L}")
    print(f'\n   file      "{file_meta.get("name")}"  {chiave_file}')
    print(f"   nodo      {node_id}  {doc.get('type')}  ·  {ancoraggio['variantCount']} varianti")
    print(f"   modificato {file_meta.get('lastModified')}")
    totale = sum(len(p["docFrames"]) for p in per_pagina)
    print(f"\n▸ PAGINE DEL FILE ({len(canvases)})   frame di doc trovati: {totale}")
    for p in per_pagina:
        if p["docFrames"]:
            print(f"   {len(p['docFrames']):>3} frame  {p['page']}")
    if not totale:
        print("   ⚠️  nessun frame con nomi tipo Purpose / Usage / Behavior / Anatomy")
    print(f"\n✅ scritto data/contracts/{slug}/figma.json\n{L}\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
